bundles/org.eclipse.equinox.http.servlet/src/org/eclipse/equinox/http/servlet/internal/context/ContextController.java

```diff
@@ -134,29 +134,78 @@ public ContextController(
+			trackingContext, ServletContextAttributeListener.class.getName(),
+			new ContextListenerTrackerCustomizer(
+				trackingContext, httpServiceRuntime, this));
+
+		servletContextAttributeListenerServiceTracker.open();
+
+		servletRequestListenerServiceTracker = new ServiceTracker<EventListener, AtomicReference<ListenerRegistration>>(
+			trackingContext, ServletRequestListener.class.getName(),
+			new ContextListenerTrackerCustomizer(
+				trackingContext, httpServiceRuntime, this));
+
+		servletRequestListenerServiceTracker.open();
+
+		servletRequestAttributeListenerServiceTracker = new ServiceTracker<EventListener, AtomicReference<ListenerRegistration>>(
+			trackingContext, ServletRequestAttributeListener.class.getName(),
+			new ContextListenerTrackerCustomizer(
+				trackingContext, httpServiceRuntime, this));
+
+		servletRequestAttributeListenerServiceTracker.open();
+
+		httpSessionListenerServiceTracker = new ServiceTracker<EventListener, AtomicReference<ListenerRegistration>>(
+			trackingContext, HttpSessionListener.class.getName(),
+			new ContextListenerTrackerCustomizer(
+				trackingContext, httpServiceRuntime, this));
+
+		httpSessionListenerServiceTracker.open();
+
+		httpSessionAttributeListenerServiceTracker = new ServiceTracker<EventListener, AtomicReference<ListenerRegistration>>(
+			trackingContext, HttpSessionAttributeListener.class.getName(),
+			new ContextListenerTrackerCustomizer(
+				trackingContext, httpServiceRuntime, this));
+
+		httpSessionAttributeListenerServiceTracker.open();
+
+		ServletContext servletContext = httpServiceRuntime.getParentServletContext();
+
+		if ((servletContext.getMajorVersion() >= 3) && (servletContext.getMinorVersion() > 0)) {
+			httpSessionIdListenerServiceTracker = new ServiceTracker<EventListener, AtomicReference<ListenerRegistration>>(
+				trackingContext, HttpSessionIdListener.class.getName(),
+				new ContextListenerTrackerCustomizer(
+					trackingContext, httpServiceRuntime, this));
+
+			httpSessionIdListenerServiceTracker.open();
+		}
+		else {
+			httpSessionIdListenerServiceTracker = null;
+		}
 
 		filterServiceTracker = new ServiceTracker<Filter, AtomicReference<FilterRegistration>>(
-			trackingContext, httpServiceRuntime.getFilterFilter(),
+			trackingContext, Filter.class,
 			new ContextFilterTrackerCustomizer(
 				trackingContext, httpServiceRuntime, this));
 
 		filterServiceTracker.open();
 
 		servletServiceTracker =  new ServiceTracker<Servlet, AtomicReference<ServletRegistration>>(
-			trackingContext, httpServiceRuntime.getServletFilter(),
+			trackingContext, Servlet.class,
 			new ContextServletTrackerCustomizer(
 				trackingContext, httpServiceRuntime, this));
 
 		servletServiceTracker.open();
 
 		resourceServiceTracker = new ServiceTracker<Object, AtomicReference<ResourceRegistration>>(
-			trackingContext, httpServiceRuntime.getResourceFilter(),
+			trackingContext, Object.class,
```

Comment:
@rotty3000 sorry about the Object.class, did not know there is explicit words saying service is typeless. The original intention was to force a type on the usages(which is the Object), in order to speed up the service tracker opening. Otherwise it would scan the entire registry. The testing was done on portal side, but only limitted to what we use from http whiteboard. I don't really know how to create tests on the whiteboard project. Clearly we don't want to break the spec, so this piece can be reverted back to the original filter way. However the need for liferay to have a speedy service tracker opening is still there. We can consider 3 possible options:
1) Update the spec to require a marker type for the resource service, I know this one is a long shot, but if it is doable, everyone would benefit from the faster service tracker opening.
2) Add a custimization point here to allow a forgein code to change how the service tracker is constructed.
3) Isolate this piece out and keep it as a liferay only patch change. We currently don't have any resource usages, if we do, we can require Object.class as one of its service type.
Let me know how do you prefer to handle it.
Thanks